### PR TITLE
kvserver: group fields mutated under raftMu+mu

### PR DESCRIPTION
**Pull request by pav-kv**
This PR groups fields mutated under both `raftMu+mu` into a section of `Replica` fields explicitly marked/commented as such. All the fields in this section can be read with only one of the two mutexes held.

This PR also eliminates a few cases when `Replica.mu` is locked unnecessarily.

Epic: none
Release note: none

### CODE REVIEW

**Comment by pav-kv** [comment]
@tbg I'm on the fence w.r.t. naming/placing this section. Initially I placed it in `mu.orRaftMu` sub-section, but then it looked unnecessarily verbose so I moved it to a dedicated `shMu` section.

File: pkg/kv/kvserver/replica.go [hunk 1]
```diff
@@ -436,6 +436,46 @@ type Replica struct {
 	// timestamp, independent of the source.
 	sideTransportClosedTimestamp sidetransportAccess
 
+	// shMu contains "shared" fields which are mutated while both raftMu and mu are
+	// held. They can be accessed when either of the two mutexes is held.
+	//
+	// TODO(pav-kv): audit all other fields and include here.
+	shMu struct {
```
Comment (pav-kv) [inline pkg/kv/kvserver/replica.go:443]:
```
"Do not communicate by sharing memory; instead, share memory by communicating."
```

**Comment by pav-kv** [comment]
There is apparently some deadlock in CI, might have accidentally caused by a few `Lock`s added in tests to maintain the fact that these fields must be modified while both mutexes are held.

File: pkg/kv/kvserver/replica.go [hunk 2]
```diff
@@ -1289,7 +1288,9 @@ func (r *Replica) entireSpanExcludedFromBackupRLocked(
 
 // Version returns the replica version.
 func (r *Replica) Version() roachpb.Version {
-	if r.mu.state.Version == nil {
```
Comment (tbg) [inline pkg/kv/kvserver/replica.go:1287]: oh, this was a bug, right?

File: pkg/kv/kvserver/replica_closedts_internal_test.go [hunk 3]
```diff
@@ -565,10 +565,12 @@ func TestReplicaClosedTimestamp(t *testing.T) {
 			cfg.TestingKnobs.DontCloseTimestamps = true
 			cfg.ClosedTimestampReceiver = &r
 			tc.StartWithStoreConfig(ctx, t, stopper, cfg)
+			tc.repl.raftMu.Lock()
```
Comment (tbg) [inline pkg/kv/kvserver/replica_closedts_internal_test.go:568]: Hmm, in light of the need to hold both mutexes, `mu.orRaftMu` is a deceptive name. (Disregard if this changes by the end of the commit list).

File: pkg/kv/kvserver/replica_application_result.go [hunk 4]
```diff
@@ -405,6 +405,8 @@ func (r *Replica) makeReproposal(origP *ProposalData) (reproposal *ProposalData,
 }
 
 func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *replicatedCmd) error {
```
Comment (tbg) [inline pkg/kv/kvserver/replica_application_result.go:407]: Rename with `raftMuLocked` suffix? Let's stick to the naming conventions unless there's a strong reason not to.

File: pkg/kv/kvserver/replica.go (hunk 1 above)
Comment (tbg) [reply pkg/kv/kvserver/replica.go:443]: Much better! Thank you for slogging through another mass rename. I like where you got this.

Comment (tbg) [reply pkg/kv/kvserver/replica.go:443]: Not going to review this commit line-by-line on the assumption that it's 100% mechanical - let me know if there are parts that could use a human check.

**Review by tbg — APPROVED** [review]
LGTM and thanks. Let's get this in before it rots.

File: pkg/kv/kvserver/replica.go (hunk 1 above)
Comment (pav-kv) [reply pkg/kv/kvserver/replica.go:439]: The big commits are 100% mechanical. The "interesting" ones (where I removed unnecessary locking) are separate and small, so hopefully you reviewed them.

**Comment by pav-kv** [comment]
TFTR!

bors r=tbg

**Pull request merged**